### PR TITLE
first drafts of ISx in Chapel

### PR DESCRIPTION
**Pull request by bradcray**
This merge contains two versions of the new ISx benchmark (https://github.com/ParRes/ISx) ported to Chapel.  Though more work remains to be done, these versions are mature enough to start getting correctness testing run on them.

Both versions use a reasonably explicit SPMD-style approach in Chapel -- creating a number of tasks, each of which does local histogramming and exchanging of elements with its siblings.  The difference between the two versions is that isx-spmd.chpl does SPMD over the Chapel locales (compute nodes) themselves, while the isx-bucket-spmd.chpl version introduces a 'numBuckets' abstraction and does SPMD over them.  Both versions use 'forall' loops within each bucket to make use of intra-locale parallelism, so there's no need to create multiple buckets per locale.  These forall loops use atomic operations to coordinate with one another, so if one wanted to run with numBuckets == numLocales*[numCores | maxTaskPar], it would be better to turn the forall loops into for loops and remove the atomic operations, as they'd be unnecessary.

From a communication standpoint, the inter-locale/bucket exchange step should be the primary source of data transfers if everything is working correctly.  Chapel's bulk communication optimizations should permit these communications to be done in one step per locale, but currently due to an int(32)/int(64) mismatch between the source and destination it can't.  We've decided to get the working version in and under performance testing before fixing the problem to see the performance delta in the nightly performance tracking graphs.

Some other TODOs for the future include:
* making the use of the PCG RNG more closely resemble the reference version with the goal of generating identical random numbers in our version as in the reference (statistically, it shouldn't matter as long as each is generating good RNGs, but just to make things apples-to-apples)
* introducing timers and enabling performance testing/graphing of these tests
* and other (generally more minor) TODOs are in the code

### CODE REVIEW

**Comment by bradcray** [comment]
@benharsh, @lydia-duncan : We often don't do code reviews for test changes, but I thought I'd offer you the opportunity to review this if you wanted since you've each spent some time with it.  It's OK just to rubber stamp it too, though.

File: test/studies/isx/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.good
```
Comment (lydia-duncan) [inline test/studies/isx/.gitignore]: I think we usually just maintain the top-level .gitignore file

Comment (bradcray) [inline test/studies/isx/.gitignore]: Oh, really?  That seems unfortunately centralized.  Is there a reason that that's better?

Comment (lydia-duncan) [inline test/studies/isx/.gitignore]: It allows the blanket ignoring of a.out, for instance.  I think there's some other common test files that are ignored en masse, and it keeps things in a centralized location.  I don't think it's a big deal, though.

Comment (bradcray) [inline test/studies/isx/.gitignore]: For something that we want to ignore everywhere, like a.out, it completely makes sense to only mention it in the top-level file.  For something that we can't ignore in most places (like *.good,) putting a long relative path into a top-level .gitignore seems harder to maintain than a per-directory .gitignore (which better supports moving the test around, removing the entry if the test is removed in the future, etc.)

Comment (bradcray) [inline test/studies/isx/.gitignore]: (I'll ask for opinions on chat when the current conversation dies down a bit)

Comment (bradcray) [inline test/studies/isx/.gitignore]: (For the github record: On chat, a few opinions in favor of decentralized .gitignores came in, none opposed)

File: test/studies/isx/isx-bucket-spmd.chpl [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+// The bucket width to use per bucket when running in weakISO mode
+//
+config const isoBucketWidth = if mode == scaling.weakISO then 8192 else 0;
+
+
+//
+// (Issue a warning if this has been set in modes other than weakISO)
+//
+if !quiet && mode != scaling.weakISO && isoBucketWidth != 0 then
+  warning("Note that isoBucketWidth has no effect for weakISO scaling mode");
```
Comment (lydia-duncan) [inline test/studies/isx/isx-bucket-spmd.chpl]: Isn't this warning the opposite of what your comment above says it should be?

Comment (bradcray) [inline test/studies/isx/isx-bucket-spmd.chpl]: I don't think so (unless there's something about the English wording that's ambiguous and you're interpreting it differently than I am?)  If (1) the scaling mode is not weakISO and (2) the bucket width is not 0 then it implies someone has set it and a warning is issued.

It seems to work as advertised too:
    ./a.out --mode=weak --isoBucketWidth=20
    /Users/bradc/chapel/modules/internal/ChapelIO.chpl:717: warning: Note that isoBucketWidth has no effect for weakISO scaling mode
    scaling mode = weak

vs.

    ./a.out --mode=weakISO --isoBucketWidth=20
    scaling mode = weakISO

Comment (lydia-duncan) [inline test/studies/isx/isx-bucket-spmd.chpl]: The warning says "Note that isoBucketWidth has no effect for weakISO scaling mode", but I think it should say "Note that isoBucketWidth only impacts weakISO scaling mode".  The first one implies to me that it *shouldn't* be used with weakISO mode.

Comment (bradcray) [inline test/studies/isx/isx-bucket-spmd.chpl]: Oh, now I see!  I thought you were saying that the warning wasn't generated when the comment said it was.  It's actually that the warning message is incorrect.  I agree, thanks for catching that.

Comment (lydia-duncan) [inline test/studies/isx/isx-bucket-spmd.chpl]: Sorry :) I'm usually more explicit with my wording nits, it's easy to just see what you thought you wrote when pointed at it generally

**Pull request merged**